src/transactions.rs

```diff
@@ -21,35 +28,284 @@ pub const RBF_SEQUENCE: u32 = u32::MAX - 2;
 /// using the new_*() methods.
 pub trait RevaultTransaction: fmt::Debug + Clone + PartialEq {
     /// Get the inner transaction
-    fn inner_tx(&self) -> &Transaction;
+    fn inner_tx(&self) -> &Psbt;
 
     /// Get the inner transaction
-    fn inner_tx_mut(&mut self) -> &mut Transaction;
+    fn inner_tx_mut(&mut self) -> &mut Psbt;
+
+    /// Add a signature in order to eventually satisfy this input.
+    /// Some sanity checks against the PSBT Input are done here, but no signature check.
+    /// The BIP174 Signer.
+    fn add_signature(
+        &mut self,
+        input_index: usize,
+        pubkey: bitcoin::PublicKey,
+        rawsig: Vec<u8>,
+    ) -> Result<Option<Vec<u8>>, Error> {
+        if let Some(ref mut psbtin) = self.inner_tx_mut().inputs.get_mut(input_index) {
+            // BIP174:
+            // For a Signer to only produce valid signatures for what it expects to sign, it must
+            // check that the following conditions are true:
+            // -- If a witness UTXO is provided, no non-witness signature may be created.
+            if psbtin.witness_utxo.is_none() {
+                return Err(Error::InputSatisfaction(format!(
+                    "No previous witness txo for psbtin: '{:?}'",
+                    psbtin
+                )));
+            }
+            if psbtin.non_witness_utxo.is_some() {
+                return Err(Error::InputSatisfaction(format!(
+                    "Unexpected non-witness txo for psbtin: '{:?}'",
+                    psbtin
+                )));
+            }
+
+            // -- If a witnessScript is provided, the scriptPubKey or the redeemScript must be for
+            // that witnessScript
+            if let Some(witness_script) = &psbtin.witness_script {
+                let expected_script_pubkey =
+                    bitcoin::Address::p2wsh(witness_script, bitcoin::Network::Bitcoin)
+                        .script_pubkey();
+                if expected_script_pubkey
+                    != psbtin
+                        .witness_utxo
+                        .as_ref()
+                        .expect("Just check it's not none")
+                        .script_pubkey
+                {
+                    return Err(Error::InputSatisfaction(format!(
+                        "Invalid witness script of previous txo ScriptPubKey for psbtin: '{:?}'",
+                        psbtin
+                    )));
+                }
+            }
+
+            // -- If a sighash type is provided, the signer must check that the sighash is acceptable.
+            // If unacceptable, they must fail.
+            if let Some(sighash_type) = rawsig.last().copied() {
```

Comment:
I may be wrong, but what if rawsig has no sighash_type and last() returns last byte of the sig ?
Is .last() the best way to retrieve if exists, the sighash_type

Reply:
I may be confused by the comment `If a sighash type is provided..`

Reply:
Yeah, this assumes valid signature. This should probably not. At least i can type the rawsig parameter to take a secp256k1 sig and bitcoin::SighashType tuple.

But should we verify the sig here ? Hmmm.
My main reason for not adding it here was that i don't want us to rely on this contextless verification downstream. OTOH we don't get the sighash from our peer, instead we create it so we may have enough context here..

Sorry for the useless comment, thinking out loud here!

Reply:
@JSwambo wdyt ?

Reply:
> I may be confused by the comment If a sighash type is provided..

Copied it straight from the BIP, which advises to check signatures here and which i purposefully did not do for the aforementioned reason. Now i'm wondering if checking the sighash but not the sig makes sense..

Reply:
Went for the `BitcoinSig` type but no sig check, see fixup commit.